state.go

```diff
@@ -0,0 +1,110 @@
+package frankenphp
+
+import (
+	"slices"
+	"strconv"
+	"sync"
+)
+
+type stateID int
+
+const (
+	// livecycle states of a thread
+	stateBooting stateID = iota
+	stateInactive
+	stateReady
+	stateShuttingDown
+	stateDone
+
+	// states necessary for restarting workers
+	stateRestarting
+	stateYielding
+
+	// states necessary for transitioning between different handlers
+	stateTransitionRequested
+	stateTransitionInProgress
+	stateTransitionComplete
+)
+
+type threadState struct {
+	currentState stateID
+	mu           sync.RWMutex
+	subscribers  []stateSubscriber
+}
+
+type stateSubscriber struct {
+	states []stateID
+	ch     chan struct{}
+}
+
+func newThreadState() *threadState {
+	return &threadState{
+		currentState: stateBooting,
+		subscribers:  []stateSubscriber{},
+		mu:           sync.RWMutex{},
+	}
+}
+
+func (ts *threadState) is(state stateID) bool {
+	ts.mu.RLock()
+	defer ts.mu.RUnlock()
+	return ts.currentState == state
+}
+
+func (ts *threadState) compareAndSwap(compareTo stateID, swapTo stateID) bool {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	if ts.currentState == compareTo {
+		ts.currentState = swapTo
+		return true
+	}
+	return false
+}
```

Comment:
```suggestion
func (ts *threadState) compareAndSwap(compareTo stateID, swapTo stateID) bool {
	ts.mu.Lock()
	ok := ts.currentState == compareTo
	if ok {
		ts.currentState = swapTo
	}
	ts.mu.Unlock()

	return ok
}
```

Reply:
I actually also realized that thread subscribers are not informed in this case and fixed it (will add a test)